### DISCUSSION TITLE
SX1276 and DTMF tones

### DISCUSSION THREAD

**Discussion by D-Alexandr** [jgromes/RadioLib#878] [Q&A]
Dear RadioLib Team,

I am using your library to work with the SX1276 radio module. I wanted to share my interest and inquire about the capabilities of your library regarding working with DTMF tones through the SX1276 module.

My current project involves receiving DTMF signals through the radio module, and I am curious about the possibility of working with these signals within the context of your library. Specifically, I would like to know if there is a way to receive any tone signal and then work with them.

Any information you can provide about the capabilities or recommendations regarding this matter would be extremely helpful for my work.

I appreciate your work on this library and the time you dedicate to responding to user inquiries. If there is any additional information or details that would be helpful to address this question, please let me know.

**Comment by jgromes** [comment]
Hi, RadioLib team here.

I don't think it's going to be very easy. Digital transceivers like SX127x are just not meant to be used like this. In the past, I was able to get them to transmit simple tones to be used by RTTY or SSTV. However, so far it has all been single tones. That's because the "tones" are generated by feeding a square wave to the SX127x. The transceiver can only take a digital input, that's why it needs to be a square wave. As you can see, that approach will not work once we need to transmit two tones at the same time.

There might still be a way to achieve this, though I have no idea if this would actually work. I was able to playback a WAV file via SX1278 and my SDR was able to receive it, so maybe DTMF would also work ...? I used the TMRpcm library (https://github.com/TMRh20/TMRpcm/tree/master), though unfortunately I don't have the source code anymore. I do have a video though: https://www.youtube.com/watch?v=FAT3iFEyLvE

While transmission could maybe perhaps be possible, I don't think reception is possible even in theory. I have no idea how you would get the tones from the digital/square wave output. There's an active issue #569 for APRS reception, which is order of magnitude easier in my opinion (because you are only dealing with single tone at a time), and I still wasn't able to do it.

**Reply by D-Alexandr** [reply]
I am currently working with radio modules and have been utilizing the SI4432 module, which offers a unique feature where it can trigger an interrupt based on a technical "prefix" received on a particular frequency. This functionality allows for the identification of the beginning of a data transmission, simplifying the process of handling valid data packets.

The "prefix" acts as a marker for the start of a transmission, and while its reception does not guarantee the integrity of the subsequent data, it serves as an indicator for the module to anticipate incoming data. Moreover, I have observed that capturing only the "prefix" without the accompanying data yields a clearer response.

For instance, the module's wake-up mode is set to detect the "0b10101010" prefix at a frequency of 1.75kHz (Call signal for ham radio). In a narrow-band FM transmission, a tone of 1.75kHz translates to a continuous stream of "10101010" (for GFSK modulation, with a 10kHz step, a logic 0 represents the carrier frequency "X," and a logic 1 represents the carrier frequency "X+10kHz"). The module then selects the first received signal that meets the criteria and generates an interrupt signal to the microcontroller pin. If the interrupt is configured only for this event, the code execution becomes streamlined.

I am curious to know if the SX1276 module offers a similar functionality or if there are any comparable features that allow for the detection of specific packages "prefixes" to initiate data reception

**Reply by jgromes** [reply]
I think what you mean by DTMF is not what I mean by DTMF. I thought you were asking about [dual-tone multi-frequency signalling](https://en.wikipedia.org/wiki/Dual-tone_multi-frequency_signaling#Keypad), where different signals are composed of two tones transmitted simultaneously. To the best of my knowledge, that is not possible on digital transceivers supported by RadioLib for reasons I mentioned above.

What you describe seem to be just detection of a single 1750 Hz tone. 

> a tone of 1.75kHz translates to a continuous stream of "10101010"

Logic suggests that would depend on the GFSK bitrate, no?

> The "prefix" acts as a marker for the start of a transmission

Do you mean the preamble? Or the sync word? As far as I know Si443x has no "pattern detected" interrupt, only a preamble or sync word detected interrupt. I'm just trying to figure out what exactly is this functionality so that I can tell you if the SX127x has something similar.

**Reply by D-Alexandr** [reply]
>Logic suggests that would depend on the GFSK bitrate, no?
 
Yes.
> Do you mean the preamble? Or the sync word?

Yes.

DTMF would be cool, but tone detection for MCU activation would also work.

My goal is to detect a call from a ham radio and transmit a tone in response. It's something similar to a beacon for fox hunting but with remote activation. The logic is as follows:
The SX1276 receives a 1750Hz tone (call signal) and triggers an interrupt for the MCU. Then, the MCU generates a response, such as three serial tones or Morse code

**Reply by jgromes** [reply]
> Do you mean the preamble? Or the sync word?

> Yes.

Well ... which one then? Preamble or sync word?

SX127x does have the ability to trigger an interrupt on preamble, it's quite common feature. RadioLib doesn't implement it directly though, so you would have to enable low-level or god mode and edit the registers manually.

You could also take a look at the [direct receive example](https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_Receive_Direct/SX127x_Receive_Direct.ino), which shows how to use the direct mode to receive a full FSK packet. You could use it to just detect your tone as a preamble by configuring the proper bit rate.

**Reply by D-Alexandr** [reply]
God mode is the next one in the plans😏  Can I somehow see what preamble code comes to me and then set it?

**Reply by jgromes** [reply]
Preamble has no "code", it's just a 101010... sequence. In the example I linked above, the program waits for a 16-bit long preamble and then a 2-byte sync word because that's the expected packet structure. In your case, you can wait for just the preamble e.g. by calling `radio.setDirectSyncWord(0x5555, 16);` or whatever lenbgth is appropriate for your bit rate and expected tone length.

**Reply by D-Alexandr** [reply]
Hi, I finally achieved a great result using your example for[ channel activity detection](https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_Channel_Activity_Detection_Receive/SX127x_Channel_Activity_Detection_Receive.ino). I'm using the preamble as a trigger for the tone burst frequency used by many ham radio. At the moment, I believe this is the maximum achievable for tone reception, so my issue is probably resolved. Thank you for your help!

**Reply by jgromes** [reply]
CAD is intended to detect LoRa preambles, if it works then that's likely an accident, probably just because the signal is very strong, and you won't be able to differentiate between different tone frequencies. Seems it would be more reliable to use direct reception ...

**Reply by D-Alexandr** [reply]
I tested direct reception at different bit rate and different bandwidths, but the module was still unable to detect the tone(1, 1.75, 2 kHz)
